scripts/modules/b2b-account/account-hierarchy.js

```diff
@@ -59,8 +61,44 @@ define(["modules/jquery-mozu", 'modules/api', "underscore", "hyprlive", "modules
             var accountId = e.currentTarget.dataset.mzValue;
         },
         changeParentAccount: function (e) {
-            //todo: need to implement this method in future.
             var accountId = e.currentTarget.dataset.mzValue;
+            var self = this;
+            inValidParentAccounts = [];
+            var validParentAccounts = [];
```

Comment:
Reviewing this method I have a few suggestions:
1. You don't need the `inValidParentAccounts` variable. All the extra pushing and looping is inefficient.
2. All the logic to generate the list of `supportedParentAccounts` should be in a single method called `getSupportedParentAccounts(currentAccount, supportedParentAccounts)` that takes the current account and the list of parent accounts from the add child account story (named `userChildAccounts` but that is a misleading name).
3. Inside `getValidParentAccounts` you can remove the current account and current parent account and then recursively loop through and remove any accounts that are children/descendants.
```
var filteredParentAccounts = supportedParentAccounts.slice() // copy the array

// Filter out the current account and parent account
filteredParentAccounts = filteredParentAccounts .filter(function (account) {
    return account.id != currentAccount.id && account.id != currentAccount.parentAccountId;
}

// Recursively filter out all accounts that are children or descendants of the current account
// TODO: implement the recursive filter here

return filteredParentAccounts;
```

Reply:
Filter logic didn't work for grandchildren because we filter out the parents and then we don't find parents for grandchildren.
so refactor existing logic and created small methods.